Sources/DeepLearning/Operators/Basic.swift

```diff
@@ -0,0 +1,303 @@
+                Tensor<Int32>(rangeFrom: start, to: result.rank, stride: 1)])
+            return result.transposed(withPermutations: resultPermutation)
+        }
+
+        let castedShape = Tensor<I>(shapeTensor)
+        var batchIndices = indices
+        var accumulated = Tensor<I>(ones: [])
+        for d in (1...batchDims).reversed() {
+            accumulated *= castedShape[d]
+            let dValue = castedShape[d - 1]
+            let dIndices = Tensor<I>(
+                rangeFrom: Tensor<I>(zeros: []),
+                to: dValue,
+                stride: Tensor<I>(ones: [])
+            ) * accumulated
+            let dShape = Tensor<Int32>(d - 1).stacked(with: [
+                Tensor<Int32>(dValue), 
+                Tensor<Int32>(indices.rank - 1)])
+            batchIndices += dIndices.reshaped(toShape: dShape)
+        }
+
+        let flatIndices = batchIndices.flattened()
+        let outerShape = shapeTensor[Int(batchDims + 1)...]
+        let innerShape = shapeTensor[..<Int(batchDims + 1)].product(squeezingAxes: [0])
+        let flatTensor = reshaped(toShape: innerShape.rankLifted().concatenated(with: outerShape))
+        let flatResult = flatTensor.gathering(atIndices: flatIndices)
+        return flatResult.reshaped(toShape: indices.shapeTensor.concatenated(with: outerShape))
+    }
+
+    /// Applies the provided boolean mask to this tensor.
+    ///
+    /// For example:
+    /// ```
+    /// // 1-D example
+    /// // tensor is [0, 1, 2, 3]
+    /// // mask is [true, false, true, false]
+    /// tensor.masked(with: mask) // is [0, 2]
+    /// 
+    /// // 2-D example
+    /// // tensor is [[1, 2], [3, 4], [5, 6]]
+    /// // mask is [true, false, true]
+    /// tensor.masked(with: mask) // is [[1, 2], [5, 6]]
+    /// ```
+    ///
+    /// In general, `0 < mask.rank = K <= tensor.rank`, and the `mask`'s shape must match the first 
+    /// K dimensions of the `tensor`'s shape. We then have:
+    /// `tensor.masked(with: mask)[i, j1, ..., jd] = tensor[i1, ..., iK, j1, ..., jd]`, where 
+    /// `[i1, ..., iK]` is the `i`th `true` entry of `mask` (row-major order).
+    /// 
+    /// The `axis` could be used with `mask` to indicate the axis to mask from. In that case, 
+    /// `axis + mask.rank <= tensor.rank` and the `mask``'s shape must match the first 
+    /// `axis + mask.rank` dimensions of the `tensor`'s shape.
+    /// 
+    /// - Parameters:
+    ///   - mask: K-D boolean tensor, where `K <= self.rank`.
+    ///   - axis: 0-D integer tensor representing the axis in `self` to mask from, where 
+    ///     `K + axis <= self.rank`.
+    /// 
+    /// - Precondition: The `mask` cannot be a scalar: `mask.rank != 0`.
+    /// 
+    /// - Returns: `(self.rank - K + 1)`-dimensional tensor populated by entries in this tensor 
+    ///   corresponding to `true` values in `mask`.
+    @inlinable
+    func masked(with mask: Tensor<Bool>, alongAxis axis: Int32 = 0) -> Tensor {
```

Comment:
Although TensorFlow calls it `boolean_mask`, I think it's a bit misleading in that it changes the shape of the tensor. I see it as another form of `gathering`. Seems better to call it `gathering(where:alongAxis:)`.